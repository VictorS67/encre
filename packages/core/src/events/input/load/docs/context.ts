/**
 * Defines the structure for inputs used to create a readable context. This interface is for feeding page
 * content and associated metadata into processing workflows.
 *
 * @template Metadata The type for metadata associated with the context, allowing for custom data structures.
 */
export interface ContextInput<Metadata = Record<string, unknown>> {
  /**
   * The main content of the page, typically as a string.
   */
  pageContent: string;

  /**
   * Optional metadata associated with the page content, providing additional information or annotations.
   */
  metadata?: Metadata;
}

/**
 * Represents a single readable context derived from a page of content. This class encapsulates the content
 * and metadata, providing a structured way to handle such information within processing or analytical operations.
 *
 * @template Metadata The type for metadata that augments the page content with additional data or annotations.
 */
export class Context<Metadata = Record<string, unknown>>
  implements ContextInput<Metadata>
{
  /**
   * The content of the page stored as a string.
   */
  pageContent = '';

  /**
   * Metadata associated with the page content, of type Metadata.
   */
  metadata: Metadata;

  /**
   * Constructs a new instance of Context, initializing it with specified page content and optional metadata.
   * @param fields An object containing pageContent and optionally metadata, conforming to ContextInput.
   */
  constructor(fields: ContextInput<Metadata>) {
    this.pageContent = fields.pageContent
      ? fields.pageContent.toString()
      : this.pageContent;
    this.metadata = fields.metadata ?? ({} as Metadata);
  }

  /**
   * Static method to check if a given value is an instance of Context.
   * @param value The value to check.
   * @returns True if the value is an instance of Context, false otherwise.
   */
  static isContext(value: unknown): value is Context {
    return value instanceof Context;
  }
}

/**
 * Represents types that can be treated as a Context. This includes single Context instances,
 * arrays of Context, or strings that could potentially be converted into a Context.
 */
export type ContextLike = Context | Context[] | string;

/**
 * Determines whether a given value qualifies as a ContextLike. This function is used to validate
 * if a value can either be directly used as a Context, converted to one, or is an array of Contexts.
 *
 * @param value The value to evaluate.
 * @returns True if the value is a string, a Context instance, or an array of Context instances; false otherwise.
 */
export function isContextLike(value: unknown): value is ContextLike {
  if (Array.isArray(value)) {
    return value.every((v) => Context.isContext(v));
  }

  return typeof value === 'string' || Context.isContext(value);
}
