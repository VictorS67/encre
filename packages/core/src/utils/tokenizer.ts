import {
  Tiktoken,
  TiktokenBPE,
  TiktokenEncoding,
  TiktokenModel,
  getEncodingNameForModel,
} from 'js-tiktoken/lite';
import { AsyncCaller } from './asyncCaller.js';

const cache: Record<string, Promise<TiktokenBPE>> = {};

const caller = new AsyncCaller({});

export async function getNumTokens(
  text: string,
  model: string
) {
  // fallback for calulating text's tokens
  // 1 token ~= 4 chars in English
  if (text.length === 0) {
    return 0;
  }

  let numTokens: number = Math.ceil(text.length / 4);

  let token: Tiktoken | undefined;
  try {
    token = await encodingForModel(getTiktokenModel(model));
  } catch (e) {
    console.warn(
      `Failed to calculate correct number of tokens, now we use 1 token ~= 4 chars in English for approximate token calculation: ${e}`
    );
  }

  if (token) {
    numTokens = token.encode(text).length;
  }

  return numTokens;
}

export async function encodingForModel(
  model: TiktokenModel,
  options?: {
    signal?: AbortSignal;
    extendedSpecialTokens?: Record<string, number>;
  }
): Promise<Tiktoken> {
  const encoding: TiktokenEncoding = getEncodingNameForModel(model);

  return getEncoding(encoding, options);
}

export async function getEncoding(
  encoding: TiktokenEncoding,
  options?: {
    signal?: AbortSignal;
    extendedSpecialTokens?: Record<string, number>;
  }
): Promise<Tiktoken> {
  if (!(encoding in cache)) {
    cache[encoding] = caller
      .fetch(`https://tiktoken.pages.dev/js/${encoding}.json`, {
        signal: options?.signal,
      })
      .then((res) => res.json() as unknown as TiktokenBPE)
      .catch((e) => {
        delete cache[encoding];
        throw e;
      });
  }

  return new Tiktoken(await cache[encoding], options?.extendedSpecialTokens);
}

export function getTiktokenModel(modelName: string): TiktokenModel {
  if (modelName.startsWith('gpt-3.5-turbo-16k')) {
    return 'gpt-3.5-turbo-16k';
  }

  if (modelName.startsWith('gpt-3.5-turbo')) {
    return 'gpt-3.5-turbo';
  }

  if (modelName.startsWith('gpt-4-32k')) {
    return 'gpt-4-32k';
  }

  if (modelName.startsWith('gpt-4')) {
    return 'gpt-4';
  }

  return modelName as TiktokenModel;
}
