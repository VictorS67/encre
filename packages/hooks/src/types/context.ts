export interface Context {
  pageContent: string;
  metadata?: Record<string, unknown>;
}
