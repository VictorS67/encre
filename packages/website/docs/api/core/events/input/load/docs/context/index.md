# events/input/load/docs/context

## Index

### Classes

- [Context](classes/Context.md)

### Interfaces

- [ContextInput](interfaces/ContextInput.md)

### Type Aliases

- [ContextLike](type-aliases/ContextLike.md)

### Functions

- [isContextLike](functions/isContextLike.md)
