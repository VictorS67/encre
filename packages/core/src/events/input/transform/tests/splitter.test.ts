import fs from 'node:fs';
import path from 'path';
import url from 'url';
import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import { stringify } from 'yaml';
import { getNumTokens } from '../../../../utils/tokenizer.js';
import { Context } from '../../load/docs/context.js';
import {
  ContextSplitter,
  RecursiveTextSplitter,
  TextSplitter,
  TokenTextSplitter,
} from '../splitter.js';

test('test custom ContextSplitter', async () => {
  class TestSplitter extends ContextSplitter {
    async split(text: string): Promise<string[]> {
      return text.split('');
    }
  }

  const testSplitter = new TestSplitter();

  const serializedStr: string = JSON.stringify(testSplitter, null, 2);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  const context = new Context({
    pageContent: 'a\nb\nc',
  });

  const newContexts = await testSplitter.invoke(context);
  expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
    'a',
    '\n',
    'b',
    '\n',
    'c',
  ]);
  expect(newContexts.map((c) => c.metadata)).toStrictEqual([
    { loc: { lines: { from: 1, to: 1 } } },
    { loc: { lines: { from: 1, to: 2 } } },
    { loc: { lines: { from: 2, to: 2 } } },
    { loc: { lines: { from: 2, to: 3 } } },
    { loc: { lines: { from: 3, to: 3 } } },
  ]);
});

test('test TextSplitter', async () => {
  // The TextSplitter does not leverage maxSize and overlap attributes.
  // It only splits the given text based on the separator.
  // The default keepSeparator is false, which means the separator is removed
  // in splitted context.
  // If the splitted context has content length that is larger than the
  // maxSize, then the warning will show up.

  const textSplitter = new TextSplitter({
    maxSize: 3,
    overlap: 1,
    separator: '+++',
    keepSeparator: false,
  });

  const serializedStr: string = JSON.stringify(textSplitter, null, 2);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  const context = new Context({
    pageContent: 'abcd+++b+++c',
  });

  jest.spyOn(global.console, 'warn');
  const newContexts = await textSplitter.invoke(context);

  expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
    'abcd',
    'b',
    'c',
  ]);
  expect(newContexts.map((c) => c.metadata)).toStrictEqual([
    { loc: { lines: { from: 1, to: 1 } } },
    { loc: { lines: { from: 1, to: 1 } } },
    { loc: { lines: { from: 1, to: 1 } } },
  ]);

  expect(console.warn).toHaveBeenCalledTimes(1);
  expect(console.warn).toHaveBeenLastCalledWith(
    `Created a chunk with a size of ${
      'abcd'.length
    } which is larger than the max size of ${textSplitter.maxSize}`
  );
});

test('test TextSplitter with separator', async () => {
  // If the keepSeparator is true, then the separator is kept in
  // splitted context.

  const textSplitter = new TextSplitter({
    maxSize: 3,
    overlap: 1,
    separator: '+++',
    keepSeparator: true,
  });

  const context = new Context({
    pageContent: 'abcd+++b+++c',
  });

  const newContexts = await textSplitter.invoke(context);

  expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
    'abcd',
    '+++b',
    '+++c',
  ]);
  expect(newContexts.map((c) => c.metadata)).toStrictEqual([
    { loc: { lines: { from: 1, to: 1 } } },
    { loc: { lines: { from: 1, to: 1 } } },
    { loc: { lines: { from: 1, to: 1 } } },
  ]);
});

test('test TextSplitter with newlines separator', async () => {
  // If the separator is newlines, then the separator is removed
  // in the splitted context. The reason for this is that the
  // splitted context is trimed before returning.

  const textSplitter = new TextSplitter({
    maxSize: 3,
    overlap: 1,
    separator: '\n\n\n',
    keepSeparator: true,
  });

  const serializedStr: string = JSON.stringify(textSplitter, null, 2);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  const context = new Context({
    pageContent: 'abcd\n\n\nb\n\n\nc',
  });

  const newContexts = await textSplitter.invoke(context);

  expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
    'abcd',
    'b',
    'c',
  ]);
  expect(newContexts.map((c) => c.metadata)).toStrictEqual([
    { loc: { lines: { from: 1, to: 1 } } },
    { loc: { lines: { from: 1, to: 1 } } },
    { loc: { lines: { from: 1, to: 1 } } },
  ]);
});

describe('RecursiveTextSplitter', () => {
  let splitter: RecursiveTextSplitter;

  beforeEach(() => {
    splitter = new RecursiveTextSplitter({
      maxSize: 3,
      overlap: 1,
      separators: ['\n\n', '\n', ' ', ''],
      keepSeparator: true,
    });
  });

  test('should split a simple text without needing recursion', async () => {
    const context = new Context({
      pageContent: 'abcd',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toStrictEqual(['abc', 'cd']);
    expect(newContexts.map((c) => c.metadata)).toStrictEqual([
      { loc: { lines: { from: 1, to: 1 } } },
      { loc: { lines: { from: 1, to: 1 } } },
    ]);
  });

  test('should return the entire text as one segment if under maxSize', async () => {
    const context = new Context({
      pageContent: 'ab',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toStrictEqual(['ab']);
    expect(newContexts.map((c) => c.metadata)).toStrictEqual([
      { loc: { lines: { from: 1, to: 1 } } },
    ]);
  });

  test('should handle an empty string input', async () => {
    const context = new Context({
      pageContent: '',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toStrictEqual([]);
    expect(newContexts.map((c) => c.metadata)).toStrictEqual([]);
  });

  test('should correctly use the first applicable separator', async () => {
    const context = new Context({
      pageContent: 'ab\n\ncd',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toStrictEqual(['ab', 'cd']);
    expect(newContexts.map((c) => c.metadata)).toStrictEqual([
      { loc: { lines: { from: 1, to: 1 } } },
      { loc: { lines: { from: 3, to: 3 } } },
    ]);
  });

  test('should dynamically remove higher granularity separators after each split', async () => {
    const context = new Context({
      pageContent: 'ab\ncd\n\nef',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
      'ab',
      'cd',
      'ef',
    ]);
    expect(newContexts.map((c) => c.metadata)).toStrictEqual([
      { loc: { lines: { from: 1, to: 1 } } },
      { loc: { lines: { from: 2, to: 2 } } },
      { loc: { lines: { from: 4, to: 4 } } },
    ]);
  });

  test('should recursively split text using nested separators', async () => {
    const context = new Context({
      pageContent: 'abcd\ne\n\nfg\nh',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
      'abc',
      'cd',
      'e',
      'fg',
      'h',
    ]);
    expect(newContexts.map((c) => c.metadata)).toStrictEqual([
      { loc: { lines: { from: 1, to: 1 } } },
      { loc: { lines: { from: 1, to: 1 } } },
      { loc: { lines: { from: 2, to: 2 } } },
      { loc: { lines: { from: 4, to: 4 } } },
      { loc: { lines: { from: 5, to: 5 } } },
    ]);
  });

  test('should ensure that all final splits conform to the size constraints', async () => {
    const context = new Context({
      pageContent: 'abcdefg',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).not.toContain('abcdefg');
  });

  test('should identify and accumulate "good splits"', async () => {
    const context = new Context({
      pageContent: 'a b cd efgh',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toContain('a b');
  });

  test('should merge accumulated "good splits" correctly', async () => {
    splitter.maxSize = 3;
    const context = new Context({
      pageContent: 'a b cd efgh',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
      'a b',
      'cd',
      'ef',
      'fgh',
    ]);
  });

  test('should handle edge cases where splits are borderline maxSize', async () => {
    splitter.maxSize = 6;
    const context = new Context({
      pageContent: 'a b cd efgh',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
      'a b cd',
      'efgh',
    ]);
  });

  test('should correctly identify and concatenate overlaps to maintain context', async () => {
    splitter.maxSize = 4;
    splitter.overlap = 2;
    const context = new Context({
      pageContent: 'a b c d e f',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
      'a b',
      'b c',
      'c d',
      'd e',
      'e f',
    ]);
  });

  test('should handle texts with no natural splitting points', async () => {
    const context = new Context({
      pageContent: 'abcde',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toStrictEqual(['abc', 'cde']);
  });

  test('should maintain performance efficiency with large texts', async () => {
    const filePath: string = path.resolve(
      path.dirname(url.fileURLToPath(import.meta.url)),
      './examples/long_text.txt'
    );

    const fdr: string = fs.readFileSync(filePath, {
      encoding: 'utf8',
      flag: 'r',
    });

    const context = new Context({
      pageContent: fdr,
    });

    const startTime = Date.now();
    await splitter.invoke(context);
    const endTime = Date.now();
    const duration = endTime - startTime;
    expect(duration).toBeLessThan(1000);
  });

  test('should maintain performance efficiency with large texts for token counting', async () => {
    // The TextSplitter allows to change the function for computing the context size.
    // Here we are using the gpt2 encoding to split the codes. However, it takes longer time
    // since we are fetching for tokens.
    // Thus, we are using the maximum context size for gpt2 model (which is 2048) to limit the
    // context size of the splitter, which is a better fit for the real-world cases.

    splitter.computeContextSize = async (text: string): Promise<number> => {
      return getNumTokens(text, 'gpt2');
    };
    splitter.maxSize = 2048;

    const filePath: string = path.resolve(
      path.dirname(url.fileURLToPath(import.meta.url)),
      './examples/long_text.txt'
    );

    const fdr: string = fs.readFileSync(filePath, {
      encoding: 'utf8',
      flag: 'r',
    });

    const context = new Context({
      pageContent: fdr,
    });

    const startTime = Date.now();
    await splitter.invoke(context);
    const endTime = Date.now();
    const duration = endTime - startTime;
    expect(duration).toBeLessThan(5000);
  });

  test('should maintain performance efficiency with large texts for token counting', async () => {
    // This is the case when the splitted context that has context size exceeded the maxSize.
    // This is because that the computContextSize first convert the text to tokens then get the
    // length of the tokens. However, the computContextSize may get different token sizes when
    // dealing with truncated context.

    splitter.computeContextSize = async (text: string): Promise<number> => {
      return getNumTokens(text, 'gpt2');
    };
    splitter.maxSize = 3;

    const context = new Context({
      pageContent: 'abcd\ne\n\nfg\nh',
    });

    const newContexts = await splitter.invoke(context);
    expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
      'abcd',
      'e',
      'fg',
      'h',
    ]);
    expect(newContexts.map((c) => c.metadata)).toStrictEqual([
      { loc: { lines: { from: 1, to: 1 } } },
      { loc: { lines: { from: 2, to: 2 } } },
      { loc: { lines: { from: 4, to: 4 } } },
      { loc: { lines: { from: 5, to: 5 } } },
    ]);
  });
});

describe('TokenTextSplitter', () => {
  let splitter: TokenTextSplitter;

  beforeEach(() => {
    splitter = new TokenTextSplitter({
      modelName: 'gpt2',
    });
  });

  test('should correctly tokenize a simple string', async () => {
    splitter.maxSize = 2;
    splitter.overlap = 1;

    const context = new Context({
      pageContent: 'Hello world!',
    });

    const newContexts = await splitter.invoke(context);

    expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
      'Hello world',
      ' world!',
      '!',
    ]);
  });

  test('should correctly handle large text based on maxSize and overlap', async () => {
    splitter.maxSize = 5;
    splitter.overlap = 2;
    const context = new Context({
      pageContent: 'This is a long test string for testing.',
    });

    const newContexts = await splitter.invoke(context);

    expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
      'This is a long test',
      ' long test string for testing',
      ' for testing.',
    ]);
  });

  test('should handle special characters based on allowed and disallowed configurations', async () => {
    splitter.maxSize = 2;
    splitter.overlap = 1;
    splitter.allowedSpecial = ['!'];
    splitter.disallowedSpecial = [];

    const context = new Context({
      pageContent: 'Hello! World?',
    });

    const newContexts = await splitter.invoke(context);

    expect(newContexts.map((c) => c.pageContent)).toStrictEqual([
      'Hello!',
      '! World',
      ' World?',
      '?',
    ]);
  });

  test('should handle empty string input', async () => {
    splitter.maxSize = 2;
    splitter.overlap = 1;

    const context = new Context({
      pageContent: '',
    });

    const newContexts = await splitter.invoke(context);

    expect(newContexts.map((c) => c.pageContent)).toStrictEqual([]);
  });
});