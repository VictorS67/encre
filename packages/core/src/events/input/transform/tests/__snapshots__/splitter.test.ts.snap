// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test TextSplitter 1`] = `
"_grp: 1
_type: constructor
_id:
  - events
  - input
  - transform
  - splitter
  - TextSplitter
_kwargs:
  max_size: 3
  overlap: 1
  separator: +++
  keep_separator: false
"
`;

exports[`test TextSplitter with newlines separator 1`] = `
"_grp: 1
_type: constructor
_id:
  - events
  - input
  - transform
  - splitter
  - TextSplitter
_kwargs:
  max_size: 3
  overlap: 1
  separator: "\\n\\n\\n"
  keep_separator: true
"
`;

exports[`test custom ContextSplitter 1`] = `
"_grp: 1
_type: constructor
_id:
  - events
  - input
  - transform
  - splitter
  - TestSplitter
_kwargs: {}
"
`;
